sdk/go/common/util/executable/executable.go

```diff
@@ -0,0 +1,51 @@
+package executable
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/sdk/go/common/util/logging"
+)
+
+const unableToFindProgramTemplate = "unable to find program: %s"
+
+// FindExecutable attempts to find the needed executable in various locations on the
+// filesystem, eventually resorting to searching in $PATH.
+func FindExecutable(program string) (string, error) {
+	if runtime.GOOS == "windows" && !strings.HasSuffix(program, ".exe") {
+		program = fmt.Sprintf("%s.exe", program)
+	}
+	// look in the same directory
+	cwd, err := os.Getwd()
+	if err != nil {
+		return "", errors.Wrap(err, "unable to get current working directory")
+	}
+
+	cwdProgram := filepath.Join(cwd, program)
+	if fileInfo, err := os.Stat(cwdProgram); !os.IsNotExist(err) && !fileInfo.Mode().IsDir() {
+		logging.V(5).Infof("program %s found in CWD", program)
+		return cwdProgram, nil
+	}
+
+	// look in $GOPATH/bin
+	if goPath := os.Getenv("GOPATH"); len(goPath) > 0 {
+		goPathProgram := filepath.Join(goPath, "bin", program)
+		if fileInfo, err := os.Stat(goPathProgram); !os.IsNotExist(err) && !fileInfo.Mode().IsDir() {
+			logging.V(5).Infof("program %s found in $GOPATH/bin", program)
+			return goPathProgram, nil
+		}
+	}
+
+	// look in the $PATH somewhere
+	if fullPath, err := exec.LookPath(program); err == nil {
```

Comment:
For the use case that's new in this PR - do we really want to be looking in CWD or `$GOPATH/bin` for `go` itself?  I would have though we could have just directly used `exec.LookPath` for that.  It appears that we do just that for finding `npm`/`yarn` as well in the equivalent codepath there.  The goal of `FindExecutable` is different - about finding user binaries.

Not a big deal though.

Reply:
I agree we could probably change this. I'm not really concerned as we've been using this same code to find the go binary from within the go language host for quite some time (`go run`, more recently `go list`) and haven't seen any issues.